src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Parser/SbomFileParser.cs

```diff
@@ -0,0 +1,314 @@
+        }
+
+        try
+        {
+            // If the end of the array is reached, return with null value to signal end of the array.
+            if (reader.TokenType == JsonTokenType.EndArray)
+            {
+                sbomFile = null;
+                return 0;
+            }
+
+            // Read the start { of this object.
+            ParserUtils.SkipNoneTokens(stream, ref buffer, ref reader);
+            ParserUtils.AssertTokenType(stream, ref reader, JsonTokenType.StartObject);
+
+            // Move to the first property name token.
+            ParserUtils.Read(stream, ref buffer, ref reader);
+            ParserUtils.AssertTokenType(stream, ref reader, JsonTokenType.PropertyName);
+
+            while (reader.TokenType != JsonTokenType.EndObject)
+            {
+                ParseProperty(ref reader, ref buffer);
+                
+                // Read the end } of this object or the next property name.
+                ParserUtils.Read(stream, ref buffer, ref reader);
+            }
+
+            // Validate the created object
+            ValidateSbomFile(this.sbomFile);
+
+            sbomFile = this.sbomFile;
+            return reader.BytesConsumed;
+        }
+        catch (EndOfStreamException)
+        {
+            sbomFile = null;
+            return 0;
+        }
+        catch (JsonException e)
+        {
+            sbomFile = null;
+            throw new ParserError($"Error while parsing JSON, addtional details: ${e.Message}", e);
+        }
+    }
+
+    private void ValidateSbomFile(SBOMFile sbomFile)
+    {
+        // I want to use the DataAnnotations Validator here, but will check with CB first
+        // before adding a new dependency.
+        var missingProps = new List<string>();
+       
+        if (sbomFile.Checksum == null || sbomFile.Checksum.Where(c => c.Algorithm == AlgorithmName.SHA256).Count() == 0)
+        {
+            missingProps.Add(nameof(sbomFile.Checksum));
+        }
+
+        if (string.IsNullOrEmpty(sbomFile.Path))
+        {
+            missingProps.Add(nameof(sbomFile.Path));
+        }
+
+        if (string.IsNullOrEmpty(sbomFile.Id))
+        {
+            missingProps.Add(nameof(sbomFile.Id));
+        }
+
+        if (string.IsNullOrEmpty(sbomFile.FileCopyrightText))
+        {
+            missingProps.Add(nameof(sbomFile.FileCopyrightText));
+        }
+
+        if (string.IsNullOrEmpty(sbomFile.LicenseConcluded))
+        {
+            missingProps.Add(nameof(sbomFile.LicenseConcluded));
+        }
+
+        if (sbomFile.LicenseInfoInFiles == null || sbomFile.LicenseInfoInFiles.Count == 0)
+        {
+            missingProps.Add(nameof(sbomFile.LicenseInfoInFiles));
+        }
+
+        if (missingProps.Count() > 0)
+        {
+            throw new ParserError($"Missing required value(s) for file object at position {stream.Position}: {string.Join(",", missingProps)}");
```

Comment:
I personally tend to prefer returning something like `ParserResult` and avoid throwing until the last possible moment.  It makes testing somewhat easier and given that all of these scenarios are expected, it's hard to classify them as truly "exceptional."  Just a thought.

Reply:
hmm.. I like that as well, might be faster as well, since throwing and catching is expensive. Did you think of something like a Monad IResult type?

Reply:
lets merge this PR as is, and I will send out a new PR with the results type in it. That will simplify some other stuff as well.